src/index.ts

```diff
@@ -0,0 +1,70 @@
+import React from 'react'
+import shallowEqual from './shallowEqual'
+
+type StateListener<T> = (state: T) => void
+type StateSelector<T, U> = (state: T) => U
+type PartialState<T> = Partial<T> | ((state: T) => Partial<T>)
+
+export default function create<
+  State extends Record<string, any>,
+  SetState extends (partialState: PartialState<Record<string, any>>) => void,
+  GetState extends () => Record<string, any>
+>(createState: (set: SetState, get: GetState) => State) {
+  const listeners: Set<StateListener<State>> = new Set()
+
+  const setState = (partialState: PartialState<State>) => {
+    state = Object.assign(
+      {},
+      state,
+      typeof partialState === 'function' ? partialState(state) : partialState
+    )
+    listeners.forEach(listener => listener(state))
+  }
+
+  const getState = () => state
+
+  const subscribe = (listener: StateListener<State>) => {
+    listeners.add(listener)
+    return () => {
+      listeners.delete(listener)
+    }
+  }
+
+  const destroy = () => {
+    listeners.clear()
+    state = {} as State
+  }
+
+  function useStore(): State
+  function useStore<U>(selector: StateSelector<State, U>): U
+  function useStore<U>(selector?: StateSelector<State, U>) {
+    // Gets entire state if no selector was passed in
+    const selectState = typeof selector === 'function' ? selector : getState
+    // Using functional initial b/c selected itself could be a function
+    const [stateSlice, setStateSlice] = React.useState(() => selectState(state))
+    // Prevent useEffect from needing to run when values change by storing them in a ref object
+    const refs = React.useRef({ stateSlice, selectState }).current
+
+    if (refs.stateSlice !== stateSlice) refs.stateSlice = stateSlice
+    if (refs.selectState !== selectState) refs.selectState = selectState
```

Comment:
Mutating refs during render is discouraged, because any render can be interrupted by "time slicing" and "suspense" features that aren't released yet.

https://twitter.com/dan_abramov/status/1102182582362521600

Reply:
useEffect(() => (functionRef.current = fn), [fn])

Would work : )

Reply:
@aleclarson Thanks for sharing this. Here's my initial thoughts:

The issue is that we shouldn’t depend on the render phase to update a ref. Why is this bad? In the case of `useInterval`, you don't want a delayed or interrupted render to cause your callback to not be updated. So is @gaearon saying that `useEffect` is not delayed or interrupted by Suspense and Time Slicing? That seems great but I would like to verify this is true.

In the case of zustand, `useStore` subscribes to the store by adding a state setter to a list. The values of the store are able to be updated outside of React by exposing a function that iterates over the list and calls the state setters. This means if an update to the store happens during a delay, state setters can be called when they "shouldn’t". I put "shouldn’t" in quotes because if the component needed to change its subscription, it was already rendering or set to render anyway. So calling its state setter should just cause the render to be rescheduled. I feel like whatever reconciliation happens during/after Suspense and Time Slicing could handle this.

Anyway, I could be completely wrong about these assumptions. We should just do what @drcmda suggests and put the ref update in `useEffect`.

Reply:
> In the case of `useInterval`, you don't want a delayed or interrupted render to cause your callback to not be updated. So is @gaearon saying that `useEffect` is not delayed or interrupted by Suspense and Time Slicing?

In the case of `useInterval`, you don't want your refs updated until the render has been applied to the DOM, which is what `useEffect` guarantees.

Reply:
Ah I see. Thanks.